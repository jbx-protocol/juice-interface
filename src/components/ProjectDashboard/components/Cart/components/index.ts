export * from './CartItem'
export * from './SummaryCollapsedView'
export * from './SummaryExpandedView'
