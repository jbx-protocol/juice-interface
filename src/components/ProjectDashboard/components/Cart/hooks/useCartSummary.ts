import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { NftRewardsContext } from 'contexts/NftRewards/NftRewardsContext'
import { useWallet } from 'hooks/Wallet'
import { useCallback, useContext, useMemo } from 'react'
import { formatCurrencyAmount } from 'utils/formatCurrencyAmount'

export const useCartSummary = () => {
  const { isConnected, connect } = useWallet()
  const cart = useProjectCart()
  const nftRewardsCtx = useContext(NftRewardsContext).nftRewards
  const tiers = useMemo(
    () => nftRewardsCtx.rewardTiers ?? [],
    [nftRewardsCtx.rewardTiers],
  )
  const nftRewards = useMemo(() => {
    return cart.nftRewards
      .map(nft => ({ ...tiers.find(tier => tier.id === nft.id), ...nft }))
      .filter(nft => nft !== undefined)
      .map(nft => ({
        fileUrl: nft.fileUrl,
        name: nft.name,
        quantity: nft.quantity,
        id: nft.id,
      }))
  }, [cart.nftRewards, tiers])

  const amountText = cart.totalAmount
    ? formatCurrencyAmount(cart.totalAmount)
    : undefined

  const resetCart = useCallback(() => {
    cart.dispatch({ type: 'reset' })
  }, [cart])

  const payProject = useCallback(() => {
    if (!isConnected) {
      connect()
      return
    }
    cart.dispatch({ type: 'openPayModal' })
  }, [cart, connect, isConnected])

  const showCurrencyOnCollapse = useMemo(() => {
    return cart.payAmount?.amount ? cart.payAmount.amount > 0 : false
  }, [cart.payAmount])

  return {
    amountText,
    currency: cart.totalAmount?.currency,
    nftRewards,
    walletConnected: isConnected,
    showCurrencyOnCollapse,
    resetCart,
    payProject,
  }
}
