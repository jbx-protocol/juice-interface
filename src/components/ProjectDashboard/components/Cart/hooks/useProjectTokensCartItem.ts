import {
  useProjectCart,
  useTokensPerEth,
} from 'components/ProjectDashboard/hooks'
import { useCallback } from 'react'

export const useProjectTokensCartItem = () => {
  const {
    totalAmount: payAmount,
    userIsReceivingTokens,
    dispatch,
  } = useProjectCart()
  const { receivedTickets } = useTokensPerEth(payAmount)

  const removeTokens = useCallback(
    () => dispatch({ type: 'removeTokens' }),
    [dispatch],
  )
  return {
    userIsReceivingTokens,
    receivedTickets,
    removeTokens,
  }
}
