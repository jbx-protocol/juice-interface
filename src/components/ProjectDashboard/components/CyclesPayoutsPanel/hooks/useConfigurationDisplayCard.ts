import { t } from '@lingui/macro'
import { useMemo } from 'react'

export const useConfigurationDisplayCard = (type: 'current' | 'upcoming') => {
  const title = useMemo(() => {
    if (type === 'current') {
      return t`Current cycle`
    }
    return t`Upcoming cycle`
  }, [type])

  return {
    title,
  }
}
