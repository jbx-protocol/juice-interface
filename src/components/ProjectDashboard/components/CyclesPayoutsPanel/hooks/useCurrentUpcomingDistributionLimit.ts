import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import useProjectDistributionLimit from 'hooks/v2v3/contractReader/useProjectDistributionLimit'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingDistributionLimit = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const {
    primaryETHTerminal,
    distributionLimit: currentDistributionLimit,
    distributionLimitCurrency: currentDistributionLimitCurrency,
  } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const {
    data: upcomingDistributionLimitData,
    loading: upcomingDistributionLimitLoading,
  } = useProjectDistributionLimit({
    projectId,
    configuration: upcomingFundingCycle?.configuration.toString(),
    terminal: primaryETHTerminal,
  })

  const [upcomingDistributionLimit, upcomingDistributionLimitCurrency] =
    upcomingDistributionLimitData ?? []

  if (type === 'current') {
    return {
      distributionLimit: currentDistributionLimit,
      distributionLimitCurrency: currentDistributionLimitCurrency,
      loading: false,
    }
  }
  return {
    distributionLimit: upcomingDistributionLimit,
    distributionLimitCurrency: upcomingDistributionLimitCurrency,
    loading: upcomingFundingCycleLoading || upcomingDistributionLimitLoading,
  }
}
