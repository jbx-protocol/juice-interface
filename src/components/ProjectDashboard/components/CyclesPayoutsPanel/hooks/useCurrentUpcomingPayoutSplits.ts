import {
  useProjectContext,
  useProjectMetadata,
} from 'components/ProjectDashboard/hooks'
import { ETH_PAYOUT_SPLIT_GROUP } from 'constants/splits'
import useProjectSplits from 'hooks/v2v3/contractReader/useProjectSplits'
import { useProjectUpcomingFundingCycle } from 'hooks/v2v3/contractReader/useProjectUpcomingFundingCycle'

export const useCurrentUpcomingPayoutSplits = (
  type: 'current' | 'upcoming',
) => {
  const { projectId } = useProjectMetadata()
  const { payoutSplits: currentPayoutSplits } = useProjectContext()

  const {
    data: upcomingFundingCycleData,
    loading: upcomingFundingCycleLoading,
  } = useProjectUpcomingFundingCycle({ projectId })
  const [upcomingFundingCycle] = upcomingFundingCycleData ?? []

  const { data: upcomingPayoutSplits, loading: upcomingProjectSplitsLoading } =
    useProjectSplits({
      projectId,
      splitGroup: ETH_PAYOUT_SPLIT_GROUP,
      domain: upcomingFundingCycle?.configuration?.toString(),
    })

  if (type === 'current') {
    return { splits: currentPayoutSplits, loading: false }
  }
  return {
    splits: upcomingPayoutSplits,
    loading: upcomingFundingCycleLoading || upcomingProjectSplitsLoading,
  }
}
