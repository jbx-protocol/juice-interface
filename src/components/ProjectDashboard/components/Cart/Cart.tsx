import { ChevronUpIcon } from '@heroicons/react/24/outline'
import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { useCallback, useEffect, useRef } from 'react'
import { twMerge } from 'tailwind-merge'
import { PayProjectModal } from '../PayProjectModal'
import { SummaryCollapsedView, SummaryExpandedView } from './components'

export const Cart = ({ className }: { className?: string }) => {
  const cart = useProjectCart()
  const cartRef = useRef<HTMLDivElement>(null)

  const toggleExpanded = useCallback(
    () => cart.dispatch({ type: 'toggleExpanded' }),
    [cart],
  )

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      // Ignore clicks on popup events
      if ((event.target as Element).closest('#headlessui-portal-root')) {
        return
      }

      if (
        cartRef.current &&
        event.target &&
        !cartRef.current.contains(event.target as Node)
      ) {
        if (cart.expanded) {
          toggleExpanded()
        }
      }
    }

    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [cart, toggleExpanded])

  return (
    <>
      <div
        data-testid="cart"
        className={twMerge(
          'fixed inset-x-0 bottom-0 z-20 h-full cursor-pointer items-center justify-center border-t border-grey-200 bg-white drop-shadow transition-all dark:border-slate-500 dark:bg-slate-900',
          cart.expanded ? 'max-h-[90vh] md:max-h-[435px]' : 'max-h-20',
          cart.visible ? 'flex' : 'hidden',
          className,
        )}
        ref={cartRef}
        onClick={toggleExpanded}
      >
        <div className="flex h-full w-full max-w-6xl items-center">
          {cart.expanded ? <SummaryExpandedView /> : <SummaryCollapsedView />}
          {!cart.expanded && (
            <ChevronUpIcon
              role="button"
              className={twMerge(
                'mr-4 h-8 w-8 flex-shrink-0',
                cart.expanded && 'mt-12 rotate-180 self-start',
              )}
            />
          )}
        </div>

        {cart.expanded && (
          <ChevronUpIcon
            role="button"
            className={twMerge(
              'absolute right-4 h-8 w-8 flex-shrink-0',
              cart.expanded && 'top-6 rotate-180 md:top-12',
            )}
          />
        )}
      </div>
      <PayProjectModal />
    </>
  )
}
