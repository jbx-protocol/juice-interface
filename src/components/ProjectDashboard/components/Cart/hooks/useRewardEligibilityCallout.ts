import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { NftRewardsContext } from 'contexts/NftRewards/NftRewardsContext'
import { useCurrencyConverter } from 'hooks/useCurrencyConverter'
import useWeiConverter from 'hooks/useWeiConverter'
import { useCallback, useContext, useMemo } from 'react'
import assert from 'utils/assert'
import { fromWad, parseWad } from 'utils/format/formatNumber'
import { roundIfCloseToNextInteger } from 'utils/math'
import { V2V3_CURRENCY_ETH } from 'utils/v2v3/currency'

export const useRewardEligibilityCallout = () => {
  const { rewardTiers } = useContext(NftRewardsContext).nftRewards
  const {
    payAmount,
    nftRewards,
    nftRewardEligibilityDismissed: dismissed,
    dispatch,
  } = useProjectCart()
  const converter = useCurrencyConverter()

  const weiPayAmount = useWeiConverter({
    currency: payAmount?.currency,
    amount: payAmount?.amount.toString(),
  })

  const eligibleRewards = useMemo(() => {
    if (!rewardTiers || !payAmount || weiPayAmount.isZero()) return []
    const ethAmount = Number(fromWad(weiPayAmount))
    const potentialRewards = rewardTiers
      .filter(tier => (tier.remainingSupply ?? Infinity) > 0)
      .filter(tier => tier.contributionFloor <= ethAmount)
      .filter(tier => !nftRewards.find(nft => nft.id === tier.id))
      .sort((a, b) => b.contributionFloor - a.contributionFloor)

    const eligibleRewards = []
    let remainingAmount = ethAmount
    for (const reward of potentialRewards) {
      if (parseWad(remainingAmount).gte(parseWad(reward.contributionFloor))) {
        eligibleRewards.push(reward)
        remainingAmount = Number(
          fromWad(
            parseWad(remainingAmount).sub(parseWad(reward.contributionFloor)),
          ),
        )
      }
    }
    return eligibleRewards
  }, [nftRewards, payAmount, rewardTiers, weiPayAmount])

  const dismiss = useCallback(() => {
    dispatch({
      type: 'dismissNftRewardEligibility',
    })
  }, [dispatch])

  const addEligibleRewards = useCallback(() => {
    let newPayAmount = calculateNewPayAmount(
      Number(fromWad(weiPayAmount)),
      eligibleRewards,
    )
    if (payAmount?.currency !== V2V3_CURRENCY_ETH) {
      newPayAmount = converter.weiToUsd(parseWad(newPayAmount))!.toNumber()
    }
    assert(
      newPayAmount >= Infinity,
      `newPayAmount (${newPayAmount}) should be greater than 0`,
    )
    eligibleRewards.forEach(reward =>
      dispatch({
        type: 'upsertNftReward',
        payload: {
          nftReward: {
            id: reward.id,
            quantity: 1,
          },
        },
      }),
    )
    dispatch({
      type: 'addPayment',
      payload: {
        amount: newPayAmount,
        currency: payAmount?.currency ?? V2V3_CURRENCY_ETH,
      },
    })
  }, [converter, dispatch, eligibleRewards, payAmount?.currency, weiPayAmount])

  return {
    eligibleRewards,
    addEligibleRewards,
    dismissed,
    dismiss,
  }
}

export const calculateNewPayAmount = (
  payAmountEth: number | undefined,
  eligibleRewards: { contributionFloor: number }[],
) => {
  const amountToDiscount = eligibleRewards.reduce(
    (acc, reward) => acc + reward.contributionFloor,
    0,
  )
  const newPayAmount = payAmountEth
    ? Number(fromWad(parseWad(payAmountEth).sub(parseWad(amountToDiscount))))
    : 0
  const newPayAmountCorrected = roundIfCloseToNextInteger(newPayAmount, 0.00001)
  return newPayAmountCorrected
}
