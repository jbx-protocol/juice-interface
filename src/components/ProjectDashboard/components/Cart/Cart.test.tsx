/**
 * @jest-environment jsdom
 */
import { fireEvent, render } from '@testing-library/react'
import { useProjectCart } from 'components/ProjectDashboard/hooks'
import { Cart } from './Cart'

jest.mock('components/ProjectDashboard/hooks/useProjectCart')

jest.mock('./components', () => ({
  SummaryCollapsedView: () => <div>SummaryCollapsedView</div>,
  SummaryExpandedView: () => <div>SummaryExpandedView</div>,
}))

jest.mock('../PayProjectModal', () => ({
  PayProjectModal: () => <div>PayProjectModal</div>,
}))

describe('Cart', () => {
  const DefaultUseProjectCartMock = {
    dispatch: jest.fn(),
    expanded: false,
    visible: true,
  }
  beforeEach(() => {
    ;(useProjectCart as jest.Mock).mockReturnValue(DefaultUseProjectCartMock)
    DefaultUseProjectCartMock.dispatch.mockReset()
  })

  it('renders', () => {
    const { container } = render(<Cart />)
    expect(container).toMatchSnapshot()
  })

  test('clicking the cart dispatches a toggle expand', () => {
    const { getByTestId } = render(<Cart />)
    const cart = getByTestId('cart')
    expect(cart).toHaveClass('max-h-20')
    fireEvent.click(cart)
    expect(useProjectCart().dispatch).toHaveBeenCalledWith({
      type: 'toggleExpanded',
    })
  })

  // test('clicking the unexpanded cart summary does not expand it', () => {
  //   const { getByTestId } = render(<Cart />)
  //   const cartSummary = getByTestId('cart-summary-closed-view-summary')
  //   expect(useProjectCart().dispatch).not.toHaveBeenCalled()
  //   fireEvent.click(cartSummary)
  //   expect(useProjectCart().dispatch).not.toHaveBeenCalled()
  // })

  // test('clicking the unexpanded cart total pay area does not expand it', () => {
  //   const { getByTestId } = render(<Cart />)
  //   const cartSummary = getByTestId('cart-summary-closed-view-total')
  //   expect(useProjectCart().dispatch).not.toHaveBeenCalled()
  //   fireEvent.click(cartSummary)
  //   expect(useProjectCart().dispatch).not.toHaveBeenCalled()
  // })

  // test('clicking the expanded cart summary does not collapse it', () => {
  //   ;(useProjectCart as jest.Mock).mockReturnValue({
  //     ...DefaultUseProjectCartMock,
  //     expanded: true,
  //   })
  //   const { getByTestId } = render(<Cart />)
  //   const cartSummary = getByTestId('cart-summary-open-view-summary')

  //   expect(useProjectCart().dispatch).not.toHaveBeenCalled()
  //   fireEvent.click(cartSummary)
  //   expect(useProjectCart().dispatch).not.toHaveBeenCalled()
  // })

  // test('clicking the expanded cart total pay area does not collapse it', () => {
  //   ;(useProjectCart as jest.Mock).mockReturnValue({
  //     ...DefaultUseProjectCartMock,
  //     expanded: true,
  //   })
  //   const { getByTestId } = render(<Cart />)
  //   const cartSummary = getByTestId('cart-summary-open-view-total')
  //   expect(useProjectCart().dispatch).not.toHaveBeenCalled()
  //   fireEvent.click(cartSummary)
  //   expect(useProjectCart().dispatch).not.toHaveBeenCalled()
  // })
})
