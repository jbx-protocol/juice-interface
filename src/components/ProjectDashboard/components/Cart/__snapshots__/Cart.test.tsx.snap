// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders 1`] = `
<div>
  <div
    class="fixed inset-x-0 bottom-0 z-20 h-full cursor-pointer items-center justify-center border-t border-grey-200 bg-white drop-shadow transition-all dark:border-slate-500 dark:bg-slate-900 max-h-20 flex"
    data-testid="cart"
  >
    <div
      class="flex h-full w-full max-w-6xl items-center"
    >
      <div>
        SummaryCollapsedView
      </div>
      <svg
        aria-hidden="true"
        class="mr-4 h-8 w-8 flex-shrink-0"
        fill="none"
        role="button"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M4.5 15.75l7.5-7.5 7.5 7.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </div>
  </div>
  <div>
    PayProjectModal
  </div>
</div>
`;
